### ISSUE TITLE
Low Concurrency and Contradictory Response timings from osm3s_query(CLI) and interpreter(CGI)

### ISSUE THREAD

**Issue by arungowtham** [drolbr/Overpass-API#283]
I have a local setup of overpass_api_v0.7.52, in my machine(32 core,16Gb Ram,2400MHz CPU)
I use this query for testing(`[out:json];area[name='Tamil Nadu'];node(area)[amenity='atm'];out count;`).
I trigger requests concurrently, and i observed these results:
- Using osm3s_query(i wrote a python wrapper to interact with osm3s_query file):
  **No.of threads(requests) | Total time taken(Seconds)**
   1-----------------------------------------1.685
   10----------------------------------------1.784
   50----------------------------------------5.590
  100---------------------------------------10.853
- Using interpreter:
  **No.of threads(requests) | Total time taken(Seconds)**
  1-----------------------------------------1.810
  10----------------------------------------3.422
  50----------------------------------------13.733
  100---------------------------------------25.452(4 requests fail showing response code: 504)
  
  I am not sure if the stats above are a normal thing, i.e i expect even 100 concurrent threads to complete within 1.685s(single request time), Even though calling the osm3s_query file directly seems better, it is not concurrent, as i scale up the threads count, the total time keeps increasing! Shouldn't it be constant?

**In short, my question's are:**
1. Why would i get contradictory response timing's? 
2. How could i tune interpreter to match CLI performance, are there any settings associated with it?
3. I would like to know the concurrency level designed in overpass api(i.e how much concurrent clients can a local setup respond to).

**Comment by mmd-osm** [comment]
How do you call `osm3s_query`? Can you post the exact command line, especially, if you're using `db-dir=` parameter? How does your python wrapper look like? What Apache settings do you have in place?

As your query is mostly CPU bound, you cannot expect 100 threads to finish in 1.6s - you only have 32 core to process your queries! Even 50 parallel threads is most likely already exceeding your available CPU resources.

**Comment by arungowtham** [comment]
1. I call osm3s_query by mentioning the db_dir & passing the query file as argument (`/poi/overpass_compiled/bin/osm3s_query --db-dir=/poi/DB_PLANET_UNCOMPRESSED/ --concise </path/to/query.json`).

Note: **Now this answers the reason for contradictory response timings** !(i.e mentioning the db-dir doesn't use dispatcher hence better response, So **does dispatcher overhead really cause this much impact on performance and response timings?**)

2. This python wrapper i mentioned, acts as the interpreter/cgi(but without using dispatcher). There is nothing specific about it, it just gets the query from apache then stores it in a temp file and calls the osm3s_query file with the command i mentioned above.

3. I couldn't get what apache setting you are trying to ask, but i will mention generally: I use apache event MPM module. Though i tried with default or buffed up settings, it dint have any impact in the response timings!!

## Response Timings :

- **osm3s_query:**
_No.of threads(requests) | Total time taken(Seconds)_
1-----------------------------------------1.685
5-----------------------------------------1.745
10----------------------------------------1.784
15----------------------------------------2.352
20----------------------------------------2.656
25----------------------------------------2.923
30----------------------------------------3.755

- **interpreter:**
_No.of threads(requests) | Total time taken(Seconds)_
1-----------------------------------------1.810
5-----------------------------------------1.961
10----------------------------------------3.422
15----------------------------------------5.367
20----------------------------------------5.758
25----------------------------------------7.831
30----------------------------------------9.559

### Inference:
The reason i trigger same requests parallel is to just benchmark(i.e i need to get an overview of how much requests can my server/overpass process), i might get even 500-1000 concurrent requests to overpass so i must be prepared to serve those huge number of requests concurrently!!

Though i have 32 cores available, i see only 5~10 requests are being processed concurrently(Observed as in response timing split up).
I can expect atleast 25~30 threads to process within 1.6s right? or will overpass utilize so much cpu resource that would limit concurrency level to ~10.
FYI: when i trigger 30 threads, 30 tasks are triggered that instant itself and every core is being utilized and %CPU is at around ~60%.

**Can i assume from your comment that, overpass is working normally in my machine** and to get further improvements/concurrency i need to scale up the specifications or opt for more servers? Or **is it just me who has mis-configured overpass to work abnormal**.

Thanks a ton for replying in such short notice @mmd-osm  :)

**Comment by mmd-osm** [comment]
I have one more question. We already talked about the 1ms wait time in the dispatcher:

http://permalink.gmane.org/gmane.comp.gis.openstreetmap.overpass/227

Did you also try this out for `interpreter`? Does it have any noticeable impact on the response time?

**Comment by arungowtham** [comment]
Yeah i tried it out, it dint have any noticeable impact on the response timings.

**Comment by mmd-osm** [comment]
(Note to self:    settings.cc: `max_num_processes(20),`   would have to be increased)

**Comment by mmd-osm** [comment]
This issue can probably be closed, follow is in https://github.com/drolbr/Overpass-API/issues/340
